src/newgrf_profiling.cpp

```diff
@@ -0,0 +1,97 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+ /** @file newgrf_profiling.cpp Profiling of NewGRF action 2 handling. */
+
+#include "newgrf_profiling.h"
+#include "date_func.h"
+#include "fileio_func.h"
+#include "string_func.h"
+#include "console_func.h"
+
+#include <chrono>
+#include <time.h>
+
+
+NewGRFProfiler *_newgrf_profiler;
+
+
+/**
+ * Create profiler object and begin profiling session.
+ * @param grffile   The GRF file to collect profiling data on
+ * @param end_date  Game date to end profiling on
+ */
+NewGRFProfiler::NewGRFProfiler(const GRFFile *grffile, Date end_date) : grffile(grffile), profile_end_date(end_date)
+{
+	IConsolePrintF(TC_LIGHT_BROWN, "Beginning profile of NewGRF %s", grffile->filename);
+}
+
+/**
+ * Complete profiling session and write data to file
+ */
+NewGRFProfiler::~NewGRFProfiler()
+{
+	std::string filename = this->GetOutputFilename();
+	IConsolePrintF(TC_LIGHT_BROWN, "Finished profile of NewGRF, writing %u events to %s", (uint)this->calls.size(), filename.c_str());
+
+	FILE *f = FioFOpenFile(filename.c_str(), "wt", Subdirectory::NO_DIRECTORY);
+	FileCloser fcloser(f);
+
+	fputs("Tick,Sprite,CallbackID,Microseconds,Subs\n", f);
```

Comment:
"GRFID", "feature" and "GRF-local" item id would be useful

Reply:
Started adding this, just not for all the various resolvers yet.